jupyter_accessibility/project.py

```diff
@@ -0,0 +1,428 @@
+    # PIP_IGNORE_INSTALLED="1",
+    PIP_NO_BUILD_ISOLATION="1",
+    # PIP_NO_DEPENDENCIES="1",
+    PYTHONIOENCODING="utf-8",
+    PYTHONUNBUFFERED="1",
+)
+
+
+class Repo(Common):
+    """the base class for all repos
+
+    repos need instructions to build jupyter products, they are defined as doit tasks in each class.
+    these repos needed to be cloned, dev dependencies built.
+    """
+
+    _EXTENSIONS = {}
+    url: str = Field(description="the url for the project pointing to a git repo")
+    branch: str = Field("main", description="the branch, ref, or hash to build")
+    path: Path = Field(None, description="path the repository with a git repo")
+    head: Path = Field(None, description="the HEAD of the git repo")
+    package: Path = Field(None, description="the package.json")
+    links: Path = Field(None, description="the yarn links location")
+    pip: str = Field(None, description="the pip command prefix including conda")
+    yarn: Path = Field(None, description="yarn prefix including conda")
+    yarn_integrity: Path = Field(None, description="the location of the yarn integrity file")
+    # useful for hashing
+    setup: Path = Field(None, description="the setup.py target")  # assuming one is being used
+
+    def __post_init__(self):
+        self.name = get_name(self.url)
+        self.dir = self.parent.dir
+        self.path = self.dir / self.name
+        self.head = self.path / ".git" / "HEAD"
+        self.package = self.path / "package.json"
+        self.links = self.path / "yarn-links"
+        self.conda = self.parent.conda
+        self.yarn = f"{self.conda} yarn --link-folder {self.links.resolve()}"
+        self.pip = f"{self.conda} python -m pip"
+        self.yarn_integrity = self.path / "node_modules" / ".yarn-integrity"
+        self.setup = self.path / "setup.py"
+
+    def get_packages(self, where="packages"):
+        """yield package.json from all folders one level down"""
+        # this requires the repos are setup/cloned
+        yield from (self.path / where).rglob("*/package.json")
+
+    # cloning requires the env
+    # everything else requires cloning
+    def clone(self, depth=1):
+        """create a task to clone a repo
+        the previous version did some more complicated things we might want to revive."""
+        # it would be a better development experience to construct the git repository
+        # explicitly relative to a commit.
+
+        # when working on local version of the repository we clone it
+        yield dict(
+            basename="setup",
+            name=self.name,
+            actions=[
+                (create_folder, [self.path]),
+                f"""{self.conda} git clone --depth {depth} --branch {self.branch} {self.url} {self.path}""",
+            ],
+            targets=[self.head],
+            clean=[f"rm -rf {self.path.absolute()}"],
+            uptodate=[self.head.exists()],
+        )
+
+    # this task needs to be generated after cloing
```

Comment:
next phase we can maybe generate the tasks as dependencies to other tasks

Reply:
yea indeed. when we do this the number of commands should become 1 rather than 5. 

when you write `doit` poorly you sometimes have to run the commands a few times until the reach a steady state. our goal should be a single command that handles the whole and parts of the workload regardless of state.